File: frontend/src/api/compliance.ts
import apiClient from './client'
import { useAuthStore } from '@/store/auth'

export type FindingStatus = 'compliant' | 'non_compliant' | 'partial' | 'na'
export type FindingSeverity = 'critical' | 'high' | 'medium' | 'low' | 'info'
export type RemediationStatus = 'open' | 'in_progress' | 'resolved' | 'accepted_risk'

export interface ComplianceFinding {
  id: string
  case_id: string
  item_id: string
  control: string
  frameworks: string // comma-joined
  title: string
  host?: string
  status: FindingStatus
  severity: FindingSeverity
  rationale: string
  remediation: string
  evidence: string
  evidence_hash?: string
  owner?: string
  due_date?: string | null
  remediation_status: RemediationStatus
  source: 'ai' | 'manual'
  created_at: string
  updated_at: string
}

export interface ComplianceSnapshot {
  id: string
  case_id: string
  total: number
  pass: number
  fail: number
  partial: number
  na: number
  score: number
  created_at: string
}

// One check sent for AI assessment.
export interface AssessItem {
  item_id: string
  control: string
  frameworks: string[]
  title: string
  purpose: string
  output: string
  host?: string
}

interface ApiResponse<T> { success: boolean; data: T }

export const complianceApi = {
  assess: async (
    caseId: string, providerId: string, items: AssessItem[], replace = true,
  ): Promise<{ assessed: number; note?: string }> => {
    const { data } = await apiClient.post<ApiResponse<{ assessed: number; note?: string }>>(
      `/cases/${caseId}/compliance/assess`, { provider_id: providerId, items, replace },
    )
    return data.data
  },

  listFindings: async (caseId: string): Promise<ComplianceFinding[]> => {
    const { data } = await apiClient.get<ApiResponse<ComplianceFinding[]>>(`/cases/${caseId}/compliance/findings`)
    return data.data
  },

  updateFinding: async (
    id: string,
    payload: {
      status?: FindingStatus; severity?: FindingSeverity; remediation?: string
      owner?: string; due_date?: string; remediation_status?: RemediationStatus
    },
  ): Promise<ComplianceFinding> => {
    const { data } = await apiClient.patch<ApiResponse<ComplianceFinding>>(`/compliance/findings/${id}`, payload)
    return data.data
  },

  listSnapshots: async (caseId: string): Promise<ComplianceSnapshot[]> => {
    const { data } = await apiClient.get<ApiResponse<ComplianceSnapshot[]>>(`/cases/${caseId}/compliance/snapshots`)
    return data.data
  },

  clearFindings: async (caseId: string): Promise<void> => {
    await apiClient.delete(`/cases/${caseId}/compliance/findings`)
  },

  // Trigger a report download (HTML) using the auth token.
  downloadReport: async (caseId: string): Promise<void> => {
    const token = useAuthStore.getState().token
    const baseURL = apiClient.defaults.baseURL ?? ''
    const res = await fetch(`${baseURL}/cases/${caseId}/compliance/report`, {
      headers: token ? { Authorization: `Bearer ${token}` } : {},
    })
    if (!res.ok) throw new Error(await res.text() || `Server ${res.status}`)
    const disposition = res.headers.get('Content-Disposition') ?? ''
    const match = disposition.match(/filename="?([^";\n]+)"?/)
    const filename = match ? match[1] : `compliance-report-${Date.now()}.html`
    const blob = await res.blob()
    const url = URL.createObjectURL(blob)
    const a = document.createElement('a')
    a.href = url
    a.download = filename
    document.body.appendChild(a)
    a.click()
    document.body.removeChild(a)
    URL.revokeObjectURL(url)
  },
}

File: backend/internal/models/compliance.go
package models

import (
	"time"

	"github.com/google/uuid"
)

// ComplianceFinding is one assessed control for a case: the result of an AI (or
// manual) evaluation of a compliance check's output against its expected
// baseline. Findings power the coverage matrix and the compliance report.
type ComplianceFinding struct {
	ID     uuid.UUID `gorm:"type:uuid;primaryKey;default:gen_random_uuid()" json:"id"`
	CaseID uuid.UUID `gorm:"type:uuid;not null;index"                       json:"case_id"`

	ItemID     string `json:"item_id"`              // checklist item id (e.g. "c-iam-w1")
	Control    string `json:"control"`              // control mapping (e.g. "ISO A.9.4.3 · PCI 8.2")
	Frameworks string `json:"frameworks"`           // comma-joined (e.g. "ISO 27001,PCI-DSS")
	Title      string `gorm:"not null" json:"title"` // the check label
	Host       string `json:"host,omitempty"`

	// Status: compliant | non_compliant | partial | na
	Status string `gorm:"default:'na'" json:"status"`
	// Severity: critical | high | medium | low | info
	Severity    string `gorm:"default:'info'" json:"severity"`
	Rationale   string `gorm:"type:text"      json:"rationale"`   // why this verdict
	Remediation string `gorm:"type:text"      json:"remediation"` // how to fix
	Evidence    string `gorm:"type:text"      json:"evidence"`     // captured output (truncated)

	// #4 Evidence integrity — SHA-256 of the evidence at assessment time.
	EvidenceHash string `json:"evidence_hash,omitempty"`

	// #6 POA&M (Plan of Action & Milestones) — remediation tracking.
	Owner             string     `json:"owner,omitempty"`
	DueDate           *time.Time `json:"due_date,omitempty"`
	RemediationStatus string     `gorm:"default:'open'" json:"remediation_status"` // open | in_progress | resolved | accepted_risk

	// Source: "ai" | "manual"
	Source    string    `gorm:"default:'ai'" json:"source"`
	CreatedBy uuid.UUID `gorm:"type:uuid"    json:"created_by"`
	CreatedAt time.Time `                    json:"created_at"`
	UpdatedAt time.Time `                    json:"updated_at"`
}

// ComplianceSnapshot captures the aggregate posture of a case at one assessment
// point, enabling drift detection (score trend across periodic audits).
type ComplianceSnapshot struct {
	ID      uuid.UUID `gorm:"type:uuid;primaryKey;default:gen_random_uuid()" json:"id"`
	CaseID  uuid.UUID `gorm:"type:uuid;not null;index"                       json:"case_id"`
	Total   int       `json:"total"`
	Pass    int       `json:"pass"`
	Fail    int       `json:"fail"`
	Partial int       `json:"partial"`
	NA      int       `json:"na"`
	Score   int       `json:"score"`
	CreatedAt time.Time `json:"created_at"`
}
